### ISSUE TITLE
COM Interop not working without dynamic keyword support

### ISSUE THREAD

**Issue by EuroEager2008** [dotnet/corefx#40053]
I reported this via Visual Studio/Send Feedback, but perhaps that is wrong place so I repeat here:
The following code works fine with full framework, but with .net core 3 (preview 7) it throws a System.IndexOutOfRangeException:

Microsoft.Office.Interop.Excel.Application excelApp = new Excel.Application();
Microsoft.Office.Interop.Excel.Workbook workbook = excelApp.Workbooks.Open(excelPath);
Microsoft.Office.Interop.Excel.Worksheet worksheet = (Microsoft.Office.Interop.Excel.Worksheet)workbook.Worksheets[“Sheet1”]; //Note the excplicit cast which is only necessary with Core, not full Fx
Microsoft.Office.Interop.Excel.Range dataRange = worksheet.Range[worksheet.Cells[2, 1], worksheet.Cells[data.GetUpperBound(0) + 2, data.GetUpperBound(1) + 1]];

The workbook (given by excelPath argument) is the exact same one in both full framework case and the core case.

**Comment by EuroEager2008** [comment]
Added info:
The following code (using dynamic and thus late binding) works fine with full Fx, but throws an exception at the last line with core:
(All dependencies to execl has been removed first)
Type appType = Type.GetTypeFromProgID("Excel.Application");
dynamic excelApp = Activator.CreateInstance(appType);
dynamic workbook = excelApp.Workbooks.Open(excelPath);

Exception:
Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: ''System.__ComObject' does not contain a definition for 'Workbooks''

I think this shows that COM interop is not really finished with preview 7, but I have no idea if it relates to my original problem though.

**Comment by jkotas** [comment]
cc @AaronRobinsonMSFT

**Comment by AaronRobinsonMSFT** [comment]
@EuroEager2008 Use of the `dynamic` keyword for COM interop is not presently supported in .NET Core - this is by design.

I will look into the issue regarding throwing `System.IndexOutOfRangeException` today.

**Comment by AaronRobinsonMSFT** [comment]
This indexer is also related to lack of `dynamic` keyword support. The problem is the CSharp binder is attempting to create an indexer out of a property, but that isn't possible in C# so that logic fails.

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/MethodOrPropertySymbol.cs#L159

A possible fix is to block the experience properly at:

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/CSharpGetMemberBinder.cs#L103-L114

In order to unblock this scenario, all implicit uses of `dynamic` would need to be avoided, but since Office PIAs default to add the `dynamic` type, this is going to be a lot of manual casting - see below.

```CSharp
Microsoft.Office.Interop.Excel.Range dataRange = worksheet.Range[
    (Microsoft.Office.Interop.Excel.Range)worksheet.Cells[2, 1],
    (Microsoft.Office.Interop.Excel.Range)worksheet.Cells[data.GetUpperBound(0) + 2, data.GetUpperBound(1) + 1]];
```

cc @jkotas @jkoritzinsky @jeffschwMSFT

**Comment by AaronRobinsonMSFT** [comment]
Closing this a deferring to https://github.com/dotnet/coreclr/issues/24246.

I have made a slight mess of this considering the real work will be in the corefx repo _not_ coreclr - i.e. Runtime binder. I am going to keep the issue in coreclr open because it has more technical details.

See also https://github.com/dotnet/corefx/issues/32630

**Comment by EuroEager2008** [comment]
Thanks
I was pretty sure I tried your work-around, but obviously not because it is working with the extra casts.

AaronRobinsonMSFT changed the milestone to Future which, I guess, means that dynamic and thereby implicit dynamic will not be introduced at Core 3 "RTM".
This makes COM interop and particularly office automation counter intuitive with .net Core and I think this will raise many support issues since it is not easy even to know where the casts are necessary to avoid dynamics (unless compiler may help which it does not currently, at least not with VS 2019 6.3 preview 1).

**Comment by AaronRobinsonMSFT** [comment]
>This makes COM interop and particularly office automation counter intuitive with .net Core and I think this will raise many support issues since it is not easy even to know where the casts are necessary to avoid dynamics

@EuroEager2008 Unfortunately that is true. Please see dotnet/coreclr#24246 which also mentions the concern. I am moving this to 5.0 for consideration. The issue is the large amount of work for porting and testing the COM dynamic binder from .NET Framework. As mentioned in the coreclr issue, porting Office automation scenarios without `dynamic` support is going to be costly and error prone. Not saying this will make 5.0, but the more feedback we receive the better the chances the cost will be acceptable.

**Comment by EuroEager2008** [comment]
Probably not well enough thought through when planning/introducing desktop additions to Core.
One thing is pretty sure: There are many .net apps out which uses Office automation and they may all be stuck with Full Fx 4.8 or something like that if this problem is not fixed at 5.0 (without major application rework).
Hopefully Office interop is more or less alone using implicit dynamics...
